basic-loadbalancer-aware-server/src/main/java/com/rtbhouse/grpc/loadbalancer/SingleLoadBalancerConnector.java

```diff
@@ -0,0 +1,335 @@
+    }
+  }
+
+  /**
+   * Sets connection (starts RPC) with the given load balancer. Waits for sending the initial report
+   * until load balancer is ready to receive it (using withWaitForReady mechanism). Once the first
+   * reply from load balancer is received, starts sending reports with the requested frequency.
+   * Callbacks (onCompleted, onError) from SignupReplyStreamObserver ensure reconnecting with load
+   * balancer in case of an error (ex. loss of network connection) or when load balancer calls
+   * onCompleted even though the server is working correctly.
+   */
+  private void connectWithLoadBalancer() {
+    logger.debug("Connecting to load balancer (starting signup rpc)...");
+    reportObserver = lbAsyncStub.withWaitForReady().signup(new SignupReplyStreamObserver());
+  }
+
+  /**
+   * Only if the connector wasn't already stopped or paused, it starts new rpc, marks the new status
+   * and sends the initial report.
+   */
+  private synchronized void reconnectWithLoadBalancer() {
+    if (status == SingleConnectorStatus.STOPPED
+        || status == SingleConnectorStatus.PAUSED
+        || status == SingleConnectorStatus.PAUSED_WHEN_CONNECTING) {
+      logger.debug("No reconnect - connector stopped, paused or paused_when_connecting");
+      return;
+    }
+
+    logger.debug("Reconnecting with load balancer...");
+    status = SingleConnectorStatus.CONNECTING;
+    connectWithLoadBalancer();
+    sendReport();
+  }
+
+  /**
+   * Prepares and sends heartbeat, or an initial report with information about the server. In the
+   * initial message includes server address, server port, list of services provided by the server
+   * and its weight, if set.
+   */
+  private void sendReport() {
+    ServerReport report;
+
+    if (status == SingleConnectorStatus.CONNECTING) { // prepare initial report
+      Server.Builder serverDetails =
+          com.rtbhouse.grpc.loadbalancer.Server.newBuilder()
+              .setIpAddress(ByteString.copyFrom(serverAddress.getAddress()))
+              .setPort(serverPort)
+              .addAllServices(Arrays.asList(services));
+
+      if (serverWeight != WEIGHT_NOT_DEFINED) {
+        serverDetails.setWeight(serverWeight);
+      }
+
+      report =
+          ServerReport.newBuilder()
+              .setReadyToServe(true)
+              .setServerDetails(serverDetails.build())
+              .build();
+
+      logger.info(
+          "Sending initial report: ready to serve true, server address {}, services {}, port {}, weight {}",
+          serverAddress.toString(),
+          String.join(", ", report.getServerDetails().getServicesList()),
+          serverPort,
+          report.getServerDetails().getWeight());
+
+    } else if (status == SingleConnectorStatus.WORKING) { // prepare heartbeat
+      report = ServerReport.newBuilder().setReadyToServe(true).build();
+      logger.debug("Sending heartbeat: ready to serve {}", report.getReadyToServe());
+
+    } else {
+      logger.warn(
```

Comment:
A jak status jest PAUSED to co? Musimy logować na WARNie?

Reply:
W sumie mogłoby być na debug, bo to tak naprawdę oznacza tyle, że sam stan zdążył się zmienić na PAUSED, ale jeszcze executor robiący heartbeaty jeszcze nie dostał .cancel(), ale nic złego się wtedy nie dzieje.